Upstream-Status: Backport

Backported from https://crrev.com/c/2049978

Signed-off-by: Maksim Sisov <msisov@igalia.com>
---
From 3f3f5e2d31cad67b92678880653d7e658496e2cb Mon Sep 17 00:00:00 2001
From: Maksim Sisov <msisov@igalia.com>
Date: Wed, 19 Feb 2020 16:56:42 +0000
Subject: [PATCH] ozone/wayland: do not use modifiers for linear buffers

Video playback is broken in Ozone/Wayland whenever Wayland compositor
announces that R_8 buffer is supported and sends supported format
modifiers for that buffer type.

Then, when video is played, GbmPixmapWayland is created with format
modifiers supplied by Wayland compositor. However, this result in a video
playback that is garbled.

The reason to this is that the buffer is requested to be linear, which
means no format modifiers should be supplied. I believe it could be
the gbm handling that, but it turns out that messes up and we have
weird behaviour.

Thus, do not use format modifiers if the buffer usage is linear.

Bug: 1050364
Change-Id: I980c462f857b3a462f95677810e3ff2aaaf340c6
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/2049978
Reviewed-by: Michael Spang <spang@chromium.org>
Reviewed-by: Robert Kroeger <rjkroege@chromium.org>
Commit-Queue: Maksim Sisov <msisov@igalia.com>
Cr-Commit-Position: refs/heads/master@{#742670}
---
 ui/ozone/platform/wayland/gpu/gbm_pixmap_wayland.cc | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/ui/ozone/platform/wayland/gpu/gbm_pixmap_wayland.cc b/ui/ozone/platform/wayland/gpu/gbm_pixmap_wayland.cc
index 737dc8251cdf..c78ca1d02aae 100644
--- a/ui/ozone/platform/wayland/gpu/gbm_pixmap_wayland.cc
+++ b/ui/ozone/platform/wayland/gpu/gbm_pixmap_wayland.cc
@@ -47,9 +47,13 @@ bool GbmPixmapWayland::InitializeBuffer(gfx::Size size,
     return false;
 
   const uint32_t fourcc_format = GetFourCCFormatFromBufferFormat(format);
-  auto modifiers = buffer_manager_->GetModifiersForBufferFormat(format);
+  auto gbm_usage = ui::BufferUsageToGbmFlags(usage);
+  std::vector<uint64_t> modifiers;
+  if (!(gbm_usage & GBM_BO_USE_LINEAR))
+    modifiers = buffer_manager_->GetModifiersForBufferFormat(format);
+
   gbm_bo_ = buffer_manager_->gbm_device()->CreateBufferWithModifiers(
-      fourcc_format, size, ui::BufferUsageToGbmFlags(usage), modifiers);
+      fourcc_format, size, gbm_usage, modifiers);
   if (!gbm_bo_) {
     LOG(ERROR) << "Cannot create bo with format= "
                << gfx::BufferFormatToString(format) << " and usage "
-- 
2.25.1

